x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -928,4 +925,154 @@ public static Expression castStringLiteral(Expression from, DataType target) {
             }
         }
     }
+
+    /**
+     * The EsqlIndexResolver will create InvalidMappedField instances for fields that are ambiguous (i.e. have multiple mappings).
+     * During ResolveRefs we do not convert these to UnresolvedAttribute instances, as we want to first determine if they can
+     * instead be handled by conversion functions within the query. This rule looks for matching conversion functions and converts
+     * those fields into MultiTypeEsField, which encapsulates the knowledge of how to convert these into a single type.
+     * This knowledge will be used later in generating the FieldExtractExec with built-in type conversion.
+     * Any fields which could not be resolved by conversion functions will be converted to UnresolvedAttribute instances in a later rule
+     * (See UnresolveUnionTypes below).
+     */
+    private static class ResolveUnionTypes extends BaseAnalyzerRule {
+
+        record TypeResolutionKey(String fieldName, DataType fieldType) {}
+
+        @Override
+        protected LogicalPlan doRule(LogicalPlan plan) {
+            List<FieldAttribute> unionFieldAttributes = new ArrayList<>();
+            // See if the eval function has an unresolved MultiTypeEsField field
+            // Replace the entire convert function with a new FieldAttribute (containing type conversion knowledge)
+            plan = plan.transformExpressionsOnly(
+                AbstractConvertFunction.class,
+                convert -> resolveConvertFunction(convert, unionFieldAttributes)
+            );
+            // If no union fields were generated, return the plan as is
+            if (unionFieldAttributes.isEmpty()) {
+                return plan;
+            }
+
+            // Otherwise drop the converted attributes after the alias function, as they are only needed for this function, and
+            // the original version of the attribute should still be seen as unconverted.
+            plan = dropConvertedAttributes(plan, unionFieldAttributes);
```

Comment:
This should not be needed - let's talk more on why this happens.

Reply:
This was the fix to the one of the issues that Luigi reported. Without this, there were multiple fields with the same name in the output, the original field with unsupported type (and null values) and the converted field (or many converted fields if we have the same field converted many times), with their converted values. But in reality all the converted fields are immediate aliased to whatever was written in the query, so the internal converted fields need to be removed.

Reply:
If I understand this well, you remove these attributes assuming they are only used in the "alias function" by which you mean an `eval x = field::type`. I don't think `eval` is the only place where you can place such a conversion function.

`STATS c=count(*) BY client_ip::ip`

`FROM sample_data, sample_data_str | SORT client_ip::ip`

Both of these have issues. After you merge this PR, I will create follow-ups for these two.